### PR TITLE
Add ConfigObserver and ConfigPropertyObserver

### LINKED ISSUE #21427: Add ConfigService observability to python
As part of PR #21221 it has become necessary to observe configuration value changes from python.
With the exposed API this is not really possible. This issue is to add a new class to the Mantid Kernel `ConfigServiceObserver` and expose it to python through `boost::python`.

### PR DESCRIPTION
**Pull request by edwardb96**
Adds two new observers `ConfigObserver` and `ConfigPropertyObserver` both of which are exposed to python. These make it possible to listen for changes to the configuration from python for changes to config property/ies.

**To test:**

Resolves #21427. 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Kernel/src/ConfigObserver.cpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#include "MantidKernel/ConfigObserver.h"
+#include "MantidKernel/ConfigService.h"
+#include <iostream>
```
Comment (martyngigg) [inline Framework/Kernel/src/ConfigObserver.cpp]: Is this needed?

File: Framework/Kernel/src/ConfigPropertyObserver.cpp [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#include "MantidKernel/ConfigPropertyObserver.h"
+#include "MantidKernel/ConfigService.h"
+#include "MantidKernel/System.h"
+#include <iostream>
```
Comment (martyngigg) [inline Framework/Kernel/src/ConfigPropertyObserver.cpp]: Is this needed?

File: Framework/Kernel/src/ConfigPropertyObserver.cpp [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#include "MantidKernel/ConfigPropertyObserver.h"
+#include "MantidKernel/ConfigService.h"
+#include "MantidKernel/System.h"
```
Comment (martyngigg) [inline Framework/Kernel/src/ConfigPropertyObserver.cpp]: This should not be necessary here.

File: Framework/Kernel/test/ConfigObserverTest.h [hunk 4]
```diff
@@ -0,0 +1,82 @@
+#ifndef MANTID_CONFIGOBSERVERTEST_H_
+#define MANTID_CONFIGOBSERVERTEST_H_
+
+#include "MantidKernel/ConfigService.h"
+#include "MantidKernel/ConfigObserver.h"
+#include "MantidKernel/System.h"
```
Comment (martyngigg) [inline Framework/Kernel/test/ConfigObserverTest.h]: This should not be necessary

File: Framework/Kernel/src/ConfigObserver.cpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace Mantid {
+namespace Kernel {
+ConfigObserver::ConfigObserver()
```
Comment (martyngigg) [inline Framework/Kernel/src/ConfigObserver.cpp]: Could we add some doxygen documentation?

File: Framework/Kernel/src/ConfigPropertyObserver.cpp [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+#include <iostream>
+
+namespace Mantid {
+namespace Kernel {
+ConfigPropertyObserver::ConfigPropertyObserver(std::string propertyName)
```
Comment (martyngigg) [inline Framework/Kernel/src/ConfigPropertyObserver.cpp]: Some doxygen documentation, please.

File: Framework/PythonInterface/test/python/mantid/kernel/ConfigPropertyObserverTest.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import unittest
+import sys
+from mantid.kernel import ConfigService, ConfigPropertyObserver
+
+if sys.version_info.major == 3:
```
Comment (martyngigg) [inline Framework/PythonInterface/test/python/mantid/kernel/ConfigPropertyObserverTest.py]: Can we swap this around and check for `2` first and then fallback: https://astrofrog.github.io/blog/2016/01/12/stop-writing-python-4-incompatible-code/

File: Framework/PythonInterface/test/python/mantid/kernel/ConfigObserverTest.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import unittest
+import sys
+from mantid.kernel import ConfigService, ConfigObserver
+
+if sys.version_info.major == 3:
```
Comment (martyngigg) [inline Framework/PythonInterface/test/python/mantid/kernel/ConfigObserverTest.py]: Can we swap this around and check for 2 first and then fallback: https://astrofrog.github.io/blog/2016/01/12/stop-writing-python-4-incompatible-code/

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: Framework/Kernel/test/ConfigObserverTest.h (hunk 4 above)
Comment (edwardb96) [reply Framework/Kernel/test/ConfigObserverTest.h]: This is required for the `UNUSED_ARG` macro which suppresses the warnings for the [unused lambda parameters](https://github.com/mantidproject/mantid/blob/432eaaacd31284e198489f7fc300cccfe31b974a/Framework/Kernel/test/ConfigObserverTest.h#L51-L57).

**Review by martyngigg — APPROVED** [review]
This is well implemented.

Good to go.

**Pull request merged**